File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.js
/**
 * @param {number[][]} graph
 * @return {number[][]}
 */

/*
depth first search
stack vs. recursion
paths = []

dfs:
if node === n - 1 ==> paths.push(path)
if !graph[node].length ==> return

for each nextNode in graph[node],
dfs(nextNode, path)
*/
var allPathsSourceTarget = function(graph) {
    const paths = [];
    
    const dfs = (node, path) => {
        if (node === graph.length - 1) {
            paths.push([...path, node]);
            return;
        }
        if (!graph[node].length) return;
        
        graph[node].forEach(nextNode => {
            dfs(nextNode, [...path, node]);
        });
    };
    
    dfs(0, []);
    return paths;
};

File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.rb
# @param {Integer[][]} graph
# @return {Integer[][]}
def dfs(node, path, graph, paths)
    if node == graph.length-1
        paths << path+[node]
        return
    end
    return if graph[node].length == 0

    graph[node].each{ |next_node| dfs(next_node, path+[node], graph, paths) }
end
def all_paths_source_target(graph)
    paths = []
    dfs(0, [], graph, paths)
    return paths
end